rfcs/web_extensions.md

```diff
@@ -0,0 +1,137 @@
+result for  a `browser.test.assert...()` method, it is up to the user agent to send the message
+"assert" or "assert-equality".
+
+Because these tests don’t leverage `testharness.js` directly, we’ve introduced a new wrapper type,
+`.extension.js`, that will create the necessary boilerplate to convert these messages into the
+corresponding assertions in the `testharness.js`.
+
+It's important to note that it isn't guaranteed that the web page will load before the extension loads.
+Meaning, a race condition could occur that would cause the `browser.test` assertion results to not
+be reported back to the test harness, since the `browser.test.onMessage` listener wasn't registered.
+To resolve this, user agents would need to queue the results until a listener is registered, and then
+report them.
+
+Last, Since the `Classic` and `BiDi` implementations support loading an extension using a `path` or
+`archivePath` to the extension's resources, the full path to the extension's resources is formulated
+by using the `test.path` before the protocol loads the extension.
+
+A proposed patch is available at https://github.com/web-platform-tests/wpt/pull/50648.
+
+##  `browser.test` API
+
+The `browser.test` API provides utilities for writing and running tests for WebExtensions.
+It includes assertion functions and a test harness to facilitate structured and reliable testing.
+
+### Methods
+
+**`browser.test.assertTrue(condition, message)`**\
+Asserts that a given condition is `true`. If the condition is `false`,
+the test fails.
+
+**Parameters**
+
+* `condition` (boolean) – The condition to assert. If `false`, the test fails.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertFalse(condition, message)`**\
+Asserts that a given condition is `false`. If the condition is `true`, the test fails.
+
+**Parameters**
+
+* `condition` (boolean) – The condition to assert. If `true`, the test fails.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertEq(actual, expected, message)`**\
+Asserts that the actual value is equal to the expected value. If not equal, the test fails.\
+This matches the ES [SameValue](https://tc39.es/ecma262/#sec-samevalue) implementation.
+
+**Parameters**
+
+* `actual` (any) – The actual value.
+* `expected` (any) – The expted value.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertThrows(fn, message)`**\
+Asserts that a given function throws an error when executed. If the function does not throw,
+the test fails.
+
+**Parameters**
+
+* `fn` (function) – A function expected to throw an error.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.runTests(tests)`**\
+Queues an array of test functions to run and returns a promise.
+The promise is rejected if one of the tests fails, otherwise resolves.
+
+**Parameters**
+
+* `tests` (array) –  An array of functions.
```

Comment:
This doesn't define how we move from one test to another, which IIRC is something we a) do want to align on and b) would require work in Chromium.  Is that in-scope to include here?

Reply:
For Safari's implementation, the tests run in the order they appear in the array. Once one test finishes, then we move onto the next test. does this match Firefox's behavior @zombie

Reply:
Yes, but see my comment below for more details.